### PR TITLE
ENH Add polymorphic has_one support to eager loading #11203

### PR DESCRIPTION
**Pull request by beerbohmdo**
## Description
Polymorphic has_one relations are currently not supported by eager-loading it just throws an error that DataObject is not a valid data class.

This PR relies on my previous PR https://github.com/silverstripe/silverstripe-framework/pull/11185, which is currently only in 5.2

## Manual testing steps

```php
<?php

use SilverStripe\ORM\DataObject;
use SilverStripe\Security\Member;
use SilverStripe\Security\Group;

/**
 * @property string|null $Title
 * @property int $PolymorphID
 * @property string $PolymorphClass
 * @method DataObject Polymorph()
 */
class MyModel extends DataObject
{
    /**
     * @var array
     * @config
     */
    private static $db = [
        'Title' => 'Varchar(255)',
    ];

    /**
     * @var array
     * @config
     */
    private static $has_one = [
        'Polymorph' => DataObject::class,
    ];
}

class MyModelController extends Controller
{
    public function index(): void
    {
        $o = MyModel::create();
        $o->Title = 'Foo';
        $o->PolymorphClass = Member::class;
        $o->PolymorphID = 1;
        $o->write();

        $o = MyModel::create();
        $o->Title = 'Bar';
        $o->PolymorphClass = Group::class;
        $o->PolymorphID = 1;
        $o->write();

        foreach (MyModel::get()->eagerLoad('Polymorph') as $item) {
            var_dump($item->Polymorph());
        }
    }
}
```

## Issues

- #11203

## Pull request checklist

- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [ ] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

### CODE REVIEW

**Comment by GuySartorelli** [comment]
Thanks for implementing this. Can you please include some manual testing steps? Preferably with some code I can just copy into a project and see it work.
That will make this a lot easier to test and therefore I'll be able to do so sooner.

**Comment by beerbohmdo** [comment]
I've added a code snippet

File: src/ORM/DataList.php [hunk 1]
```diff
@@ -1128,41 +1136,59 @@ private function fetchEagerLoadHasOne(
             if (is_array($parentData)) {
                 // $parentData represents a record in this DataList
                 $hasOneID = $parentData[$hasOneIDField];
-                $fetchedIDs[] = $hasOneID;
-                $addTo[$hasOneID][] = $parentData['ID'];
+
+                if ($hasOneID > 0) {
```
Comment (GuySartorelli) [inline src/ORM/DataList.php]: Why was this condition added? It doesn't seem related to polymorphism.

File: src/ORM/DataList.php [hunk 2]
```diff
@@ -1128,41 +1136,59 @@ private function fetchEagerLoadHasOne(
...
+                    $hasOneClass = $schema->baseDataClass($parentData[$hasOneClassField] ?? $relationDataClass);
```
Comment (GuySartorelli) [inline src/ORM/DataList.php]: Why are we getting the base data class specifically?
And why do we fall back on `$relationDataClass`?

File: src/ORM/DataList.php [hunk 3]
```diff
@@ -1128,41 +1136,59 @@ private function fetchEagerLoadHasOne(
...
-            if (isset($addTo[$fetched->ID])) {
-                foreach ($addTo[$fetched->ID] as $addHere) {
-                    if ($addHere instanceof DataObject) {
-                        $addHere->setEagerLoadedData($relationName, $fetched);
-                    } elseif (is_array($addHere)) {
-                        $addHere['list']->addEagerLoadedData($relationName, $addHere['ID'], $fetched);
-                    } else {
-                        $this->eagerLoadedData[$relationChain][$addHere][$relationName] = $fetched;
+        foreach ($fetchedIDs as $class => $ids) {
+            foreach (DataObject::get($class)->byIDs($ids) as $fetched) {
```
Comment (GuySartorelli) [inline src/ORM/DataList.php:1178]: This change increases the number of requests that are made regardless of whether the has_one is polymorphic or not.

Please refactor this to be a single query like it was before (instead of a query in a loop) or if that's not possible, only do it in a loop like this for polymorphic has_one.

File: tests/php/ORM/DataListEagerLoadingTest.php
Comment (GuySartorelli) [inline tests/php/ORM/DataListEagerLoadingTest.php:1]: Please include at least one test where the record stored in the polymorphic has_one relation is a subclass (i.e. `DataObject::getSchema()->baseDataClass()` would return the superclass)

**Review by GuySartorelli — CHANGES_REQUESTED** [review]
Awesome, thanks for this! Works well locally, just a few questions about the approach and some requested changes below.

File: src/ORM/DataList.php (hunk 1 above)
Comment (beerbohmdo) [reply src/ORM/DataList.php]: No and yes. Entries with 0 should not trigger a query. Especially for polymorph entries this is a problem because the "HasOneClass" field is mostly not empty and can contain any random class.

File: src/ORM/DataList.php (hunk 2 above)
Comment (beerbohmdo) [reply src/ORM/DataList.php]: Because the "HasOneClass" field could contain any sub class, with this will only do one fetch.

The fallback is for normal has_one relations.

But I think the fallback should be the otherway around.  `!empty(parentData[$hasOneClassField]) ? $schema->baseDataClass($parentData[$hasOneClassField]) : $relationDataClass`

File: src/ORM/DataList.php (hunk 3 above)
Comment (beerbohmdo) [reply src/ORM/DataList.php:1178]: No its not. The `$fetchedIDs` will only contain one `$class` entry for a normal has_one, but multiple for polymorph.

Its does not make sense to do a if which checks if fetchedIDs only contains one entry todo the same thing which it would do in the loop ...

File: tests/php/ORM/DataListEagerLoadingTest.php
Comment (beerbohmdo) [reply tests/php/ORM/DataListEagerLoadingTest.php:1]: I think I will remove the baseClass check completely. This may reduce the amount of queries it will trigger the first time, but you will get lazyLoad queries afterwards ...

File: src/ORM/DataList.php (hunk 2 above)
Comment (beerbohmdo) [reply src/ORM/DataList.php]: I removed the baseClass check.

The idea was to remove multiple Queries for the same base DataObject, but as it would trigger lazyLoad queries afterwards this was a bad idea.

File: src/ORM/DataList.php (hunk 1 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: Please either
1. remove these checks (as they aren't related to the problem this PR aims to fix) and open a separate issue and PR for them, OR
2. Change them to `if ($hasOneID)` which reflects `DataObject::getComponent()` and allows for people to change the implementation of primary key to allow non-integer values. Also, if there's _no_ ID, make sure we're adding a `Injector::inst()->create($class)` record (for non-polymorphic relations) or `null` (for polymorphic relations) so that this check doesn't have to happen again in `DataObject::getComponent()`.

Option 1 will make it easier and thus faster to get this bug fix reviewed and merged.

Comment (GuySartorelli) [reply src/ORM/DataList.php]: Do the same with the other identical check below.

File: src/ORM/DataList.php (hunk 3 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:1178]: Can you please try explaining that again? I don't understand what you're trying to say.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/ORM/DataList.php (hunk 3 above)
Comment (beerbohmdo) [reply src/ORM/DataList.php:1178]: With a normal `has_one` (`private static $has_one = ['Foo' => Foo::class];`), the `$fetchedIDs` array will only contain one entry (`Foo::class`). So only one query to the Database will happen, like before this change.

With a polymorphic has_one (aka. `private static $has_one = ['Foo' => DataObject::class];`) you have following database schema:

| ID | FooClass | FooID |
| --- | ------------  | -------- |
| 1 | Member | 1 |
| 2 | Member | 2 |
| 3 | Group | 1 |

To  fetch these, you need a query to Member and one to Group table.

`$fetchedIDs` will look like this:

```php
$fetchedIDs = [
   'Member' => [1, 2],
   'Group' => [1],
];
```

File: src/ORM/DataList.php (hunk 1 above)
Comment (beerbohmdo) [reply src/ORM/DataList.php]: It is related. Without this check a empty polymorphic has_one will trigger a query against random tables, because the `*Class` enum will not be empty most of the time ...

And even if it would be empty - This will end in an error.

Comment (beerbohmdo) [reply src/ORM/DataList.php]: > Also, if there's no ID, make sure we're adding a Injector::inst()->create($class) record (for non-polymorphic relations) or null (for polymorphic relations) so that this check doesn't have to happen again in DataObject::getComponent().

I could do that, but it would be a change:

```
        // NOTE: Unlike the other relation types, we don't have to explicitly fill empty DataObject records
        // into the has_one components - DataObject does that for us in getComponent() without any extra
        // db calls.
```

`null` for polymorph entries would even be a breaking change: https://github.com/silverstripe/silverstripe-framework/issues/11165


> and allows for people to change the implementation of primary key to allow non-integer values

This would be nice. But currently not really possible, many_many are explicitly created as Int for example. "by_id" only accept Int values etc.

I will change the check against empty value.

File: src/ORM/DataList.php (hunk 3 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php:1178]: Ahh, yup, that makes sense. Thank you for explaining that.

File: src/ORM/DataList.php (hunk 1 above)
Comment (GuySartorelli) [reply src/ORM/DataList.php]: > DataObject does that for us in getComponent() without any extra db calls.

I wonder if I wrote that 😂 Okay yeah if that's being handled already then I'm fine with it. The only change that's needed then is to not assume the ID is an int.

> This would be nice. But currently not really possible

It is possible, it just requires a lot of work. The more times we assume an int, the more work it will be. We should aim to reduce the number of times we assume the ID is an int rather than increasing it.

> I will change the check against empty value.

Yes please. A simple falsy `if (!$hasOneID)` is sufficient and is used elsewhere.

**Review by GuySartorelli — APPROVED** [review]
LGTM.
This should have been there already so I'm considering this a bugfix rather than an enhancement - I'll update the commit prefix on merge.
